// @agm/core/esm2015/lib/services.js
export { CircleManager } from './services/managers/circle-manager';
export { DataLayerManager } from './services/managers/data-layer-manager';
export { FitBoundsAccessor } from './services/fit-bounds';
export { AgmGeocoder } from './services/geocoder-service';
export { GoogleMapsAPIWrapper } from './services/google-maps-api-wrapper';
export { GoogleMapsScriptProtocol, LAZY_MAPS_API_CONFIG, LazyMapsAPILoader, } from './services/maps-api-loader/lazy-maps-api-loader';
export { InfoWindowManager } from './services/managers/info-window-manager';
export { KmlLayerManager } from './services/managers/kml-layer-manager';
export { LayerManager } from './services/managers/layer-manager';
export { MapsAPILoader } from './services/maps-api-loader/maps-api-loader';
export { MarkerManager } from './services/managers/marker-manager';
export { NoOpMapsAPILoader } from './services/maps-api-loader/noop-maps-api-loader';
export { PolygonManager } from './services/managers/polygon-manager';
export { PolylineManager } from './services/managers/polyline-manager';
export { RectangleManager } from './services/managers/rectangle-manager';
