// ngx-lottie/esm2015/server/lottie-server.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lottie-server.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken, APP_INITIALIZER } from '@angular/core';
import { TransferState } from '@angular/platform-browser';
import { appInitializerFactory } from './internals';
/** @type {?} */
export const LOTTIE_SERVER_OPTIONS = new InjectionToken('LottieServerOptions');
export class LottieServerModule {
    /**
     * @param {?} options
     * @return {?}
     */
    static forRoot(options) {
        return {
            ngModule: LottieServerModule,
            providers: [
                {
                    provide: LOTTIE_SERVER_OPTIONS,
                    useValue: options
                },
                {
                    provide: APP_INITIALIZER,
                    useFactory: appInitializerFactory,
                    multi: true,
                    deps: [LOTTIE_SERVER_OPTIONS, TransferState]
                }
            ]
        };
    }
}
LottieServerModule.decorators = [
    { type: NgModule }
];
