// firebaseui-angular/bundles/firebaseui-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('firebaseui'), require('@angular/fire'), require('@angular/fire/auth'), require('firebase/auth'), require('firebase/app'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('firebaseui-angular', ['exports', '@angular/core', 'firebaseui', '@angular/fire', '@angular/fire/auth', 'firebase/auth', 'firebase/app', '@angular/common'], factory) :
    (global = global || self, factory(global['firebaseui-angular'] = {}, global.ng.core, global.firebaseui$1, global.ng.fire, global.ng.fire.auth, null, global.firebaseOriginal, global.ng.common));
}(this, (function (exports, core, firebaseui$1, fire, auth, auth$1, firebaseOriginal, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    var FirebaseuiAngularLibraryService = /** @class */ (function () {
        function FirebaseuiAngularLibraryService(options, nameOrConfig, zone) {
            // noinspection JSNonASCIINames
            var app = fire["ɵfirebaseAppFactory"](options, zone, nameOrConfig);
            // store the firebaseui instance on the window object to prevent double initialization
            if (!window.firebaseUiInstance) {
                window.firebaseUiInstance = new firebaseui$1.auth.AuthUI(app.auth());
            }
            this.firebaseUiInstance = window.firebaseUiInstance;
        }
        FirebaseuiAngularLibraryService.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Inject, args: [fire.FIREBASE_OPTIONS,] }] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [fire.FIREBASE_APP_NAME,] }] },
            { type: core.NgZone }
        ]; };
        FirebaseuiAngularLibraryService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(fire.FIREBASE_OPTIONS)),
            __param(1, core.Optional()), __param(1, core.Inject(fire.FIREBASE_APP_NAME)),
            __metadata("design:paramtypes", [Object, Object, core.NgZone])
        ], FirebaseuiAngularLibraryService);
        return FirebaseuiAngularLibraryService;
    }());

    var FirebaseuiAngularLibraryComponent = /** @class */ (function () {
        function FirebaseuiAngularLibraryComponent(angularFireAuth, _firebaseUiConfig, _firebaseUiConfig_Feature, ngZone, firebaseUIService) {
            this.angularFireAuth = angularFireAuth;
            this._firebaseUiConfig = _firebaseUiConfig;
            this._firebaseUiConfig_Feature = _firebaseUiConfig_Feature;
            this.ngZone = ngZone;
            this.firebaseUIService = firebaseUIService;
            this.signInSuccessWithAuthResultCallback = new core.EventEmitter(); // tslint:disable-line
            this.signInFailureCallback = new core.EventEmitter(); // tslint:disable-line
            this.uiShownCallback = new core.EventEmitter(); // tslint:disable-line
        }
        FirebaseuiAngularLibraryComponent_1 = FirebaseuiAngularLibraryComponent;
        Object.defineProperty(FirebaseuiAngularLibraryComponent.prototype, "firebaseUiConfig", {
            get: function () {
                return __assign(__assign({}, this._firebaseUiConfig), this._firebaseUiConfig_Feature);
            },
            enumerable: true,
            configurable: true
        });
        FirebaseuiAngularLibraryComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.subscription = this.angularFireAuth.authState.subscribe(function (value) {
                if ((value && value.isAnonymous) || !value) {
                    if (_this.firebaseUiConfig.signInOptions.length !== 0) {
                        _this.firebaseUIPopup();
                    }
                    else {
                        throw new Error('There must be at least one AuthProvider.');
                    }
                }
            });
        };
        FirebaseuiAngularLibraryComponent.prototype.ngOnDestroy = function () {
            if (!!this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        FirebaseuiAngularLibraryComponent.prototype.getUIAuthConfig = function () {
            if (!this.firebaseUiConfig.callbacks) {
                this._firebaseUiConfig[FirebaseuiAngularLibraryComponent_1.COMPUTED_CALLBACKS] = true;
                this._firebaseUiConfig.callbacks = this.getCallbacks();
            }
            return this.firebaseUiConfig;
        };
        FirebaseuiAngularLibraryComponent.prototype.firebaseUIPopup = function () {
            var firebaseUiInstance = this.firebaseUIService.firebaseUiInstance;
            var uiAuthConfig = this.getUIAuthConfig();
            // Check if callbacks got computed to reset them again after providing the to firebaseui.
            // Necessary for allowing updating the firebaseui config during runtime.
            var resetCallbacks = false;
            if (uiAuthConfig[FirebaseuiAngularLibraryComponent_1.COMPUTED_CALLBACKS]) {
                resetCallbacks = true;
                delete uiAuthConfig[FirebaseuiAngularLibraryComponent_1.COMPUTED_CALLBACKS];
            }
            // show the firebaseui
            firebaseUiInstance.start('#firebaseui-auth-container', uiAuthConfig);
            if (resetCallbacks) {
                this._firebaseUiConfig.callbacks = null;
            }
        };
        FirebaseuiAngularLibraryComponent.prototype.getCallbacks = function () {
            var _this = this;
            var signInSuccessWithAuthResultCallback = function (authResult, redirectUrl) {
                _this.ngZone.run(function () {
                    _this.signInSuccessWithAuthResultCallback.emit({
                        authResult: authResult,
                        redirectUrl: redirectUrl
                    });
                });
                return _this.firebaseUiConfig.signInSuccessUrl;
            };
            var signInFailureCallback = function (error) {
                _this.ngZone.run(function () {
                    _this.signInFailureCallback.emit({
                        code: error.code,
                        credential: error.credential
                    });
                });
                return Promise.reject();
            };
            var uiShownCallback = function () {
                _this.ngZone.run(function () {
                    _this.uiShownCallback.emit();
                });
            };
            return {
                signInSuccessWithAuthResult: signInSuccessWithAuthResultCallback,
                signInFailure: signInFailureCallback,
                uiShown: uiShownCallback
            };
        };
        var FirebaseuiAngularLibraryComponent_1;
        FirebaseuiAngularLibraryComponent.COMPUTED_CALLBACKS = 'COMPUTED_CALLBACKS';
        FirebaseuiAngularLibraryComponent.ctorParameters = function () { return [
            { type: auth.AngularFireAuth },
            { type: undefined, decorators: [{ type: core.Inject, args: ['firebaseUIAuthConfig',] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: ['firebaseUIAuthConfigFeature',] }] },
            { type: core.NgZone },
            { type: FirebaseuiAngularLibraryService }
        ]; };
        __decorate([
            core.Output('signInSuccessWithAuthResult'),
            __metadata("design:type", core.EventEmitter)
        ], FirebaseuiAngularLibraryComponent.prototype, "signInSuccessWithAuthResultCallback", void 0);
        __decorate([
            core.Output('signInFailure'),
            __metadata("design:type", core.EventEmitter)
        ], FirebaseuiAngularLibraryComponent.prototype, "signInFailureCallback", void 0);
        __decorate([
            core.Output('uiShown'),
            __metadata("design:type", core.EventEmitter)
        ], FirebaseuiAngularLibraryComponent.prototype, "uiShownCallback", void 0);
        FirebaseuiAngularLibraryComponent = FirebaseuiAngularLibraryComponent_1 = __decorate([
            core.Component({
                selector: 'firebase-ui',
                template: '<div id="firebaseui-auth-container"></div>'
            }),
            __param(1, core.Inject('firebaseUIAuthConfig')),
            __param(2, core.Inject('firebaseUIAuthConfigFeature')),
            __metadata("design:paramtypes", [auth.AngularFireAuth, Object, Object, core.NgZone,
                FirebaseuiAngularLibraryService])
        ], FirebaseuiAngularLibraryComponent);
        return FirebaseuiAngularLibraryComponent;
    }());

    /*
     * Created by Raphael Jenni
     * Copyright (c) 2017 Raphael Jenni
     */
    var firebase = firebaseOriginal;
    var firebaseui = firebaseui$1;
    var FirebaseUISignInSuccessWithAuthResult = /** @class */ (function () {
        function FirebaseUISignInSuccessWithAuthResult() {
        }
        return FirebaseUISignInSuccessWithAuthResult;
    }());
    var FirebaseUISignInFailure = /** @class */ (function () {
        function FirebaseUISignInFailure() {
        }
        return FirebaseUISignInFailure;
    }());

    var FirebaseUIModule = /** @class */ (function () {
        function FirebaseUIModule() {
        }
        FirebaseUIModule_1 = FirebaseUIModule;
        FirebaseUIModule.forRoot = function (firebaseUiAuthConfig) {
            return {
                ngModule: FirebaseUIModule_1,
                providers: [
                    { provide: 'firebaseUIAuthConfig', useValue: firebaseUiAuthConfig },
                    { provide: 'firebaseUIAuthConfigFeature', useValue: {} }
                ]
            };
        };
        FirebaseUIModule.forFeature = function (firebaseUIAuthConfig) {
            return {
                ngModule: FirebaseUIModule_1,
                providers: [
                    { provide: 'firebaseUIAuthConfigFeature', useValue: firebaseUIAuthConfig }
                ]
            };
        };
        var FirebaseUIModule_1;
        FirebaseUIModule = FirebaseUIModule_1 = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule
                ],
                declarations: [FirebaseuiAngularLibraryComponent],
                providers: [FirebaseuiAngularLibraryService],
                exports: [FirebaseuiAngularLibraryComponent]
            })
        ], FirebaseUIModule);
        return FirebaseUIModule;
    }());

    exports.FirebaseUIModule = FirebaseUIModule;
    exports.FirebaseUISignInFailure = FirebaseUISignInFailure;
    exports.FirebaseUISignInSuccessWithAuthResult = FirebaseUISignInSuccessWithAuthResult;
    exports.FirebaseuiAngularLibraryComponent = FirebaseuiAngularLibraryComponent;
    exports.FirebaseuiAngularLibraryService = FirebaseuiAngularLibraryService;
    exports.firebase = firebase;
    exports.firebaseui = firebaseui;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
